docs/rules/no-access-state-in-setstate.md

```diff
@@ -0,0 +1,39 @@
+# Prevent using this.state within a this.setState (no-access-state-in-setstate)
+
+This rule should prevent usage of this.state inside setState calls.
+Such usage of this.state might result in errors when two state calls is
+called in batch and thus referencing old state and not the current
+state. An example can be an increment function:
+
+```
+function increment() {
+  this.setState({value: this.state.value + 1});
+}
+```
+
+If these two setState operations is grouped together in a batch it will
+look be something like the following, given that value is 1:
+
+```
+setState({value: 1 + 1})
+setState({value: 1 + 1})
+```
+
+This can be avoided with using callbacks which takes the previous state
+as first argument:
+
+```
+function increment() {
+  this.setState(prevState => ({value: prevState.value + 1}));
```

Comment:
Right now, I will report this as a warning. The same with 

```javascript
getstate() {
  return { value: this.state.counter++ };
}

increase() {
  this.setState(getState());
}
```

I could easily make it only report `this.state` directly inside `this.setState`, but I don't see why you would use `this.state` and not the `prevState`-solution which is guaranteed to be up-to-date.

Do you have an example when you actually would want to do that? 

This is from the [react docs](https://facebook.github.io/react/docs/react-component.html#setstate): 
> If the next state depends on the previous state, we recommend using the updater function form, instead
> ```
> this.setState((prevState) => {
>  return {counter: prevState.quantity + 1};
>}); 
> ```